src/test/java/org/opentripplanner/street/model/edge/StreetEdgeGeofencingTest.java

```diff
@@ -191,62 +230,81 @@ public void pickupFloatingVehicleWhenLeavingAZone() {
       // we want to pick up a vehicle
       final State rentalState = states[1];
       assertEquals(RENTING_FLOATING, rentalState.getVehicleRentalState());
-      assertEquals(BICYCLE, rentalState.getNonTransitMode());
+      assertEquals(BICYCLE, rentalState.currentMode());
 
       // but also keep on walking in case we don't find an edge where to leave the vehicle
       var walkingState = states[0];
       assertEquals(HAVE_RENTED, walkingState.getVehicleRentalState());
-      assertEquals(WALK, walkingState.getNonTransitMode());
+      assertEquals(WALK, walkingState.currentMode());
     }
-  }
-
-  @Test
-  public void addTwoExtensions() {
-    var edge = streetEdge(V1, V2);
-    edge.addRentalRestriction(new BusinessAreaBorder("a"));
-    edge.addRentalRestriction(new BusinessAreaBorder("b"));
 
-    assertTrue(edge.fromv.rentalTraversalBanned(forwardState("a")));
-    assertTrue(edge.fromv.rentalTraversalBanned(forwardState("b")));
-  }
-
-  @Test
-  public void removeExtensions() {
-    var edge = streetEdge(V1, V2);
-    var a = new BusinessAreaBorder("a");
-    var b = new BusinessAreaBorder("b");
-    var c = new BusinessAreaBorder("c");
+    @Test
+    public void pickupFloatingVehiclesWhenStartedInNoDropOffZone() {
+      var req = defaultArriveByRequest();
 
-    edge.addRentalRestriction(a);
+      V2.addRentalRestriction(NO_DROP_OFF_TIER);
+      V2.addRentalRestriction(noDropOffRestriction(NETWORK_BIRD));
 
-    assertTrue(edge.fromv.rentalRestrictions().traversalBanned(forwardState("a")));
+      // this is the state that starts inside a no drop off
+      // and is walking towards finding a rental vehicle
+      var haveRentedState = State
+        .getInitialStates(Set.of(V2), req)
+        .stream()
+        .filter(s -> s.getVehicleRentalState() == HAVE_RENTED)
+        .findAny()
+        .get();
 
-    edge.addRentalRestriction(b);
-    edge.addRentalRestriction(c);
+      var edge = streetEdge(V1, V2);
 
-    edge.removeRentalExtension(a);
+      var states = edge.traverse(haveRentedState);
 
-    assertTrue(edge.fromv.rentalRestrictions().traversalBanned(forwardState("b")));
-    assertTrue(edge.fromv.rentalRestrictions().traversalBanned(forwardState("c")));
+      // we return 3 states: one for the speculative renting of a vehicle, but with the information
+      // of which networks' no-drop-off zones it started in
+      assertEquals(3, states.length);
+
+      // first the speculative renting case
+      final State speculativeRenting = states[0];
+      assertEquals(RENTING_FLOATING, speculativeRenting.getVehicleRentalState());
+      assertEquals(BICYCLE, speculativeRenting.currentMode());
+      // null means that the
```

Comment:
means what?

Reply:
:) Fleshed out in 0c103bc86641e57adac694966354a860715dd228